api/replica.go

```diff
@@ -41,12 +59,34 @@ func (p *Replica) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 
 	// Otherwise, forward.
-	if p.primary == "" {
-		httpError(w, "No elected primary cluster manager", http.StatusInternalServerError)
-		return
+
+	// lock a read on primary. use primaryWait.L just to ensure we're locking
+	// and unlocking the right lock, if something chnages later
+	p.primaryWait.L.Lock()
+	// check if the primary is empty, and launch into a wait if it is
+	for p.primary == "" {
+		// give up waiting if the context is canceled, to avoid blocking forever
+		select {
+		case <-r.Context().Done():
+			httpError(w, "No elected primary cluster manager", http.StatusInternalServerError)
+			p.primaryWait.L.Unlock()
+			return
+		default:
+		}
+		// wait for the primary to become available
+		p.primaryWait.Wait()
 	}
+	primary := p.primary
+	p.primaryWait.L.Unlock()
```

Comment:
Add a comment here about how we need to release the lock before we call `hijack` because `hijack` can take an arbitrarily long time.

Reply:
i'm factoring the synchronized bits of code out into a separate method so that it's clearer and cleaner.